text/0000-never-patterns.md

```diff
@@ -0,0 +1,250 @@
+        B(..) => println!("Reachable!"),
+    }
+}
+
+// It may be possible to construct a `&!` pointing to
+// uninitialized data (unsafe, but valid).
+let never_ref: &! = ...;
+let result: Result<u8, &!> = Err(never_ref);
+match result {
+    Ok(x) => { ... }
+    // This branch would then be reached without UB.
+    Err(_) => println!("Reachable!"),
+}
+```
+
+```rust
+enum Void {}
+let result: &Result<T, Void> = ...;
+let x = match result {
+    Ok(x) => x,
+    Err(!), // No need for an arm body
+};
+// Or even
+let (Ok(x) | Err(!)) = result;
+```
+
+-->
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Patterns can be used on a partially uninitialized place; the basic rule is that a pattern only accesses data that is directly mentioned in the pattern: an `Ok(..)` pattern requires a discriminant read, a binding requires a full read, a wildcard `_` accesses nothing etc. 
+
+For uninhabited types (types with no valid values), normal patterns cannot be used to indicate an access since there is no data to be accessed. For this purpose, you can use the special `!` pattern.
+
+The `!` pattern is allowed on any uninhabited type (such as `enum Void {}` or `(Result<!, !>, u32)`) and does an access to the underlying data. Since there can be no such data (on pain of UB), the corresponding arm is known to be unreachable and does not need a body:
+```rust
+enum Void {}
+unsafe fn empty_tup<T>(tup: *const (T, Void)) -> ! {
+    unsafe {
+        match *tup { ! }
+    }
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We add `!` to the syntax of patterns. A `!` pattern is accepted for any type that is visibly uninhabited (i.e. uninhabited taking into account private fields and `#[non_exhaustive]` annotations) from the item containing the match expression.
+
+A pattern that contains a `!` is called a _never pattern_. For or-patterns, all alternatives must be never patterns for the whole pattern to be too. For example `Ok(x) | Err(!)` is not a never pattern but `(_, !)` and `Ok((_, !)) | Err(!)` are.
+
+A never pattern is accepted as an arm in a match expression, and that arm takes no body nor guards.
+
+```rust
+let x: *const Option<(u32, (u32, !))> = ...;
+match *x {
+    None => { ... }
+    Some((_, !)),
+}
+```
+
+A never pattern is also accepted in other places a pattern would be (`if let`, destructuring `let`, etc.), and makes code that corresponds to that branch unreachable.
+
+```rust
+enum Void {}
+impl Void {
+    fn ex_falso<T>(!: Self) -> T {}
+}
+```
+
+In terms of both semantics and exhaustiveness checking, a `!` pattern behaves like a binding. E.g. a `Some(!),` arm (absent match ergonomics) is semantically equivalent to `Some(x) => match x {}`. Indeed, they both indicate a load at a place of an uninhabited type, followed by an unreachable body.
+
+Never patterns interact with match ergonomics as expected: a `!` pattern is allowed on e.g. `&(T, !)`.
```

Comment:
I would have expected `&(T, !)` to be an uninhabited type, so this sentence is odd.

Reply:
Yet we haven't yet decided that it is

Reply:
If we do go with the "safety invariant" version of inhabitedness then indeed this sentence isn't useful anymore.